### ISSUE TITLE
Is there an API to check if the current controller pod is the leader in the operator cluster ?

### ISSUE THREAD

**Issue by kaushiksrinivas** [operator-framework/operator-sdk#6353]
Suppose operator cluster is run with 3 replicas in it and leader election is enabled.
One of the controller pod becomes the leader eventually.

But is there a way or api to programmatically know if the current controller is the leader or not ?
We need this api for one of the use case where each operator replica pod must be aware if it is the leader of the cluster or not.

**Comment by kaushiksrinivas** [comment]
@camilamacedo86 
Can you provide any inputs here or tag anyone who could help us ?

**Comment by OchiengEd** [comment]
What are you attempting to accomplish by detecting which controller pod is the leader and what operator-sdk version are you using? If at all you are already developing.

Only the controller pod that manages to acquire the leader election lock or lease should be running. The other pods would be basically dormant checking every so often, 15 seconds (See [LeaseDuration](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Options) ) by default.

In summary, only the leader controller pod will be reconciling events while the other two will be on standby.

**Comment by varshaprasad96** [comment]
I haven't tested it for a particular use-case, but I think this should work:

Background:
Controller-runtime internally calls in the [leader election mechanism](https://github.com/kubernetes-sigs/controller-runtime/blob/c304e7ec2ee7879f53ce69c12a7aa133853fccdb/pkg/leaderelection/leader_election.go#L101) defined by client-go. Client-go in turn [uses](https://github.com/kubernetes/client-go/blob/608b5216df92921a10d935fb9807197583f7222a/tools/leaderelection/leaderelection.go#L269-L273) "EndpointResourcesLock" which utilizes K8s leader though leases concept. Leases help in acquiring lock on shared resources by electing a leader and making sure that it is alive through heartbeats.

Long story short, `coordination.k8s.io` API Group would be available in your cluster. We could do `k get leases -A` to list out the namespaces/name and holder of the lock (The leader election name can be fetched from your [main.go](https://github.com/operator-framework/operator-controller/blob/main/main.go#L73)). In my example:

```
➜  operator-controller git:(main) ✗ k get leases -A | grep -i operatorframework
operator-controller-system   9c4404e7.operatorframework.io             operator-controller-controller-manager-5fdb4997d8-vnczp_f2307826-18d9-4cb3-ace2-53da24875c73   6d17h
```
gives me the pod ID which is currently the leader.  

This is a workaround by querying leases (which you could implement programatically), but afaik I don't think the controller-runtime [manager](https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager#Manager) provides an option to fetch the current leader ID.

**Comment by basti1302** [comment]
I realize this issue is old and closed already, but just in case someone else stumbles upon this, here is an alternative solution for operators using controller-runtime, by adding a small runnable to the manager:

In your main.go (or where ever you create the manager):
```
mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
...
})

// add the new runnable after creating the manager and before starting it
leaderElectionAwareRunnable := NewLeaderElectionAwareRunnable()
if err = mgr.Add(leaderElectionAwareRunnable); err != nil {
    return fmt.Errorf("unable to add the leader election aware runnable: %w", err)
}

if err = mgr.Start ...
```

The runnable could look like this:
```
import (
	"context"
	"sync/atomic"

	"sigs.k8s.io/controller-runtime/pkg/log"
)

type LeaderElectionAware interface {
	IsLeader() bool
}

// LeaderElectionAwareRunnable serves the purpose of making the operator manager aware whether it is the current leader
// or not.
type LeaderElectionAwareRunnable struct {
	isLeader atomic.Bool
}

func NewLeaderElectionAwareRunnable() *LeaderElectionAwareRunnable {
	return &LeaderElectionAwareRunnable{}
}

// NeedLeaderElection implements the LeaderElectionRunnable interface, which indicates
// that the LeaderElectionAwareRunnable requires leader election.
func (r *LeaderElectionAwareRunnable) NeedLeaderElection() bool {
	return true
}

// Start is the signal from controller-runtime for the runnable that this replica has become leader.
func (r *LeaderElectionAwareRunnable) Start(ctx context.Context) error {
	log.FromContext(ctx).Info("This operator manager replica has just become leader.")
	r.isLeader.Store(true)
	return nil
}

func (r *LeaderElectionAwareRunnable) IsLeader() bool {
	return r.isLeader.Load()
}
```
